internal/server/clear_task.go

```diff
@@ -0,0 +1,94 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"github.com/ministryofjustice/opg-go-common/template"
+	"github.com/ministryofjustice/opg-sirius-lpa-frontend/internal/sirius"
+	"golang.org/x/sync/errgroup"
+	"net/http"
+	"strconv"
+)
+
+type ClearTaskClient interface {
+	ClearTask(ctx sirius.Context, taskID int) error
+	Task(ctx sirius.Context, id int) (sirius.Task, error)
+}
+
+type clearTaskData struct {
+	XSRFToken string
+	Entities  []string
+	Uid       string
+	Task      sirius.Task
+	Success   bool
+	Error     sirius.ValidationError
+}
+
+func ClearTask(client ClearTaskClient, tmpl template.Template) Handler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		if err := r.ParseForm(); err != nil {
+			return err
+		}
+
+		var lpa *sirius.Case
+		var taskIDs []int
+
+		for _, id := range r.Form["id"] {
+			taskID, err := strconv.Atoi(id)
+			if err != nil {
+				return err
+			}
+			taskIDs = append(taskIDs, taskID)
+		}
+		if len(taskIDs) == 0 {
+			return errors.New("no tasks selected")
+		}
+
+		ctx := getContext(r)
+		group, groupCtx := errgroup.WithContext(ctx.Context)
+
+		data := clearTaskData{XSRFToken: ctx.XSRFToken}
+		taskID, err := strconv.Atoi(r.FormValue("id"))
+		if err != nil {
+			return err
+		}
+
+		group.Go(func() error {
+			task, err := client.Task(ctx.With(groupCtx), taskID)
+			if err != nil {
+				return err
+			}
+
+			if lpa == nil && len(task.CaseItems) > 0 {
+				lpa = &task.CaseItems[0]
+			}
+
+			data.Uid = lpa.UID
+			data.Task = task
+
+			return nil
+		})
+
+		if err := group.Wait(); err != nil {
+			return err
+		}
+
+		if r.Method == http.MethodPost {
+			err = client.ClearTask(ctx, taskID)
+
+			if ve, ok := err.(sirius.ValidationError); ok {
+				w.WriteHeader(http.StatusBadRequest)
+				data.Error = ve
+			} else if err != nil {
+				return err
+			} else if lpa != nil && lpa.CaseType == "DIGITAL_LPA" {
+				SetFlash(w, FlashNotification{Title: "Task completed"})
```

Comment:
Would it make sense if this said "Task cleared"? (I haven't seen the designs.)

Then again, if it's been through QA, I assume this was checked.

Reply:
This was checked. It will be UR'd too so future changes may happen but I think it's fine for now.